File: 1314. Matrix Block Sum/1314. Matrix Block Sum.cpp
class Solution {
public:
    vector<vector<int>> matrixBlockSum(vector<vector<int>>& mat, int K) {
        int m=mat.size(),n=mat[0].size();
        vector<vector<int>> dp(m+1,vector<int>(n+1));
        for(int i=0;i<m;++i)
            for(int j=0;j<n;++j)
                dp[i+1][j+1]=dp[i][j+1]+dp[i+1][j]-dp[i][j]+mat[i][j];
        vector<vector<int>> res(m,vector<int>(n));
        for(int i=0;i<m;++i)
            for(int j=0;j<n;++j){
                int r1=max(0,i-K),c1=max(0,j-K);
                int r2=min(m,i+K+1),c2=min(n,j+K+1);
                res[i][j]=dp[r2][c2]-dp[r2][c1]-dp[r1][c2]+dp[r1][c1];
            }
        return res;
    }
};

File: Solutions/1314. Matrix Block Sum/1314. Matrix Block Sum.py
class Solution:
    def matrixBlockSum(self, mat: List[List[int]], K: int) -> List[List[int]]:
        m,n=len(mat),len(mat[0])
        dp=[[0]*(n+1)for _ in range(m+1)]
        for i in range(m):
            for j in range (n):
                dp[i+1][j+1]=dp[i+1][j]+dp[i][j+1]-dp[i][j]+mat[i][j]
        res=[[0]*(n)for _ in range(m)]
        for i in range(m):
            for j in range(n):
                r_lo,c_lo,r_hi,c_hi=max(0,i-K),max(0,j-K),min(m,i+K+1),min(n,j+K+1)
                res[i][j]=dp[r_hi][c_hi]-dp[r_hi][c_lo]-dp[r_lo][c_hi]+dp[r_lo][c_lo]
        return res;